### PR TITLE
New module: Terraform

### PR DESCRIPTION
**Pull request by ryansb**
##### SUMMARY

New module to support creating/updating infrastructure managed by [Terraform](https://www.terraform.io/). 

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

terraform

##### ANSIBLE VERSION

```
ansible 2.5.0 (tf-module 0a52d3459d) last updated 2017/09/29 11:32:02 (GMT -400)                 
  config file = None    
  configured module search path = [u'/home/ryansb/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']      
  ansible python module location = /home/ryansb/code/ansible/lib/ansible                         
  executable location = /home/ryansb/code/ansible/bin/ansible                                    
  python version = 2.7.12 (default, Sep 27 2016, 18:08:33) [GCC 6.2.1 20160916 (Red Hat 6.2.1-2)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/cloud/misc/terraform.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+   - To just run a `terraform plan`, use check mode.
+requirements: [ "terraform" ]
+author: "Ryan Scott Brown @ryansb"
+'''
+
+EXAMPLES = """
+# Basic deploy of a service
+- terraform:
+    project_path: '{{ project_dir }}'
+    state: latest
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/misc/terraform.py]: latest -> present

File: lib/ansible/modules/cloud/misc/terraform.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+      returned: always
+      description: Whether Terraform has marked this value as sensitive
+    type:
+      type: string
+      returned: always
+      description: The type of the value (string, int, etc)
+    value:
+      returned: always
+      description: The value of the output as interpolated by Terraform
+stdout:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/misc/terraform.py:115]: Is there a way to format the stdout? When applying it gets hard to read:
```
    "sdtout": "aws_instance.example: Creating...\n  ami:                          \"\" => \"ami-ff4ea59d\"\n  associate_public_ip_address:  \"\" => \"<computed>\"\n  availability_zone:            \"\" => \"<computed>\"\n  ebs_block_device.#:           \"\" => \"<computed>\"\n  ephemeral_block_device.#:     \"\" => \"<computed>\"\n  instance_state:               \"\" => \"<computed>\"\n  instance_type:                \"\" => \"t2.micro\"\n  ipv6_address_count:           \"\" => \"<computed>\"\n  ipv6_addresses.#:             \"\" => \"<computed>\"\n  key_name:                     \"\" => \"<computed>\"\n  network_interface.#:          \"\" => \"<computed>\"\n  network_interface_id:         \"\" => \"<computed>\"\n  placement_group:              \"\" => \"<computed>\"\n  primary_network_interface_id: \"\" => \"<computed>\"\n  private_dns:                  \"\" => \"<computed>\"\n  private_ip:                   \"\" => \"<computed>\"\n  public_dns:                   \"\" => \"<computed>\"\n  public_ip:                    \"\" => \"<computed>\"\n  root_block_device.#:          \"\" => \"<computed>\"\n  security_groups.#:            \"\" => \"<computed>\"\n  source_dest_check:            \"\" => \"true\"\n  subnet_id:                    \"\" => \"<computed>\"\n  tenancy:                      \"\" => \"<computed>\"\n  volume_tags.%:                \"\" => \"<computed>\"\n  vpc_security_group_ids.#:     \"\" => \"<computed>\"\naws_instance.example: Still creating... (10s elapsed)\naws_instance.example: Still creating... (20s elapsed)\naws_instance.example: Still creating... (30s elapsed)\naws_instance.example: Creation complete after 34s (ID: i-04362e7210c113e2c)\n\nApply complete! Resources: 1 added, 0 changed, 0 destroyed.\n"
```

File: lib/ansible/modules/cloud/misc/terraform.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    command.extend(_state_args(state_file))
+
+    rc, out, err = module.run_command(command + variables_args, cwd=project_path)
+
+    if rc == 0:
+        # no changes
+        return plan_path, False
+    elif rc == 1:
+        # failure to plan
+        module.fail_json(msg='Terraform plan could not be created\r\nSTDOUT: {}\r\n\r\nSTDERR: {}'.format(out, err))
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/misc/terraform.py]: Python 2.6 needs {0}, {1} - empty brackets in various other places too.

File: lib/ansible/modules/cloud/misc/terraform.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    module = AnsibleModule(
+        argument_spec=dict(
+            project_path=dict(required=True, type='path'),
+            binary_path=dict(type='path'),
+            state=dict(default='present', choices=['present', 'absent', 'planned']),
+            variables=dict(type='dict'),
+            variables_file=dict(type='path'),
+            plan_file=dict(type='path'),
+            state_file=dict(type='path'),
+            targets=dict(type='list', default=[]),
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/misc/terraform.py:191]: targets, lock, and lock_timeout have no documentation

File: lib/ansible/modules/cloud/misc/terraform.py (hunk 2 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/misc/terraform.py:115]: Not right now, there isn't a JSON-output option for the Terraform apply run, so it'd have to be something we parse out from their output.

File: lib/ansible/modules/cloud/misc/terraform.py (hunk 3 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/misc/terraform.py]: Got em.

File: lib/ansible/modules/cloud/misc/terraform.py (hunk 4 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/misc/terraform.py:191]: Good point - added.

**Comment by thaumos** [comment]
works fine for me!  

obligatory post because @ryansb made me put this here...

**Pull request merged**